File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        def binary_search(n: int, dp: List[List[int]]) -> int:
            l, r = 0, len(dp) - 1
            
            while l < r:
                m = (l + r) // 2
                num = dp[m][-1]
                
                if num < n:
                    l = m + 1
                
                else:
                    r = m
            
            return l
        
        dp = []
        
        for n in nums:
            if not dp or dp[-1][-1] < n:
                dp.append([n])
            
            else:
                idx = binary_search(n, dp)
                dp[idx].append(n)
        
        return len(dp)

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.ts
function lengthOfLIS(nums: number[]): number {
    const dp: number[][] = []
    for (const n of nums) {
        if (!dp.length || dp.at(-1).at(-1) < n) {
            dp.push([n])
        } else {
            const idx = binarySearch(n)
            dp[idx].push(n)
        }
    }
    return dp.length
    
    function binarySearch(n: number): number {
        let l = 0, r = dp.length-1
        while (l < r) {
            const m = Math.floor((l+r)/2)
            const num = dp[m].at(-1)
            if (num < n) {
                l = m + 1
            } else {
                r = m
            }
        }
        return l
    }
};